GOSSIP.md

```diff
@@ -34,45 +35,252 @@ also used to determine which node should keep its initiated connection open in
+3. (Optionally) cache the known peers between restart
+
+To prevent an attacker from poisoning the list of peers, isolating the
+node from the rest of the network, we make it impossible to predict
+the set of peer's IP/port the attacker should control to fill enough of the list
+for the attack to be statistically successful. In addition we randomize the
+peer eviction to prevent an attacker from predicting it and replace all good
+peers for there own compromised ones.
+
+This is achieved by first categorizing peers in two pools:
+
+1. The unverified pool contains all the peers received through gossip.
+It ensures no Byzantine node can fill it completely by limiting the subset
+of the pool it can affect; it prevents attackers from predicting the set of
+IP/port they must use to successfully eclipse the node.
+
+2. The verified pool contains the peers the node connected to explicitly after
+passing through the filter of the unverified pool. It randomizes the peer
+distribution to the eye of attackers, preventing them from predicting the
+eviction algorithm.
+
+Peers are considered verified when the node has been able to connect to
+them using the Noise protocol. Incoming connections are still serviced for
+gossip but are not considered verified to prevent an attacker from circumventing
+the algorithm by connecting directly to the attacked node.
+
+Both pools clusterized peers into multiple buckets; they select which one
+an added peer belongs to based on a combination of the address of the node the
+information is coming from, the address of the peer to be added and a secret
+generated by the node for the purpose of randomizing the selection process.
+See [Unverified Pool](#unverified-pool) and [Verified Pool](#verified-pool) for
+more details on how the peers are added to the pools.
+
+The peers received from configuration are added directly to the verified pool
+and are marked as "trusted", meaning that they will never get downgraded to the
+unverified pool even if the node cannot connect to them.
+
+When started, the node will iteratively pick a random peer from the verified
+pool and try connecting to it. If there are no more peers available in the
+verified pool, a random one is picked from the unverified pool. If the
+connection succeeds, an unverified peer is upgraded to the verified pool and
+removed from the unverified pool. If the connection fails, the peer retry
+counter and last retry time are updated. The node keeps picking peers and
+connecting until the maximum number of concurrent connections is reached.
+
+The peer retry counter and last retry time (initialized to `0` and `infinity`),
+are used to filter out peers when picking them from the pools providing
+exponential backoff. After N failed attempts, a verified peer that is not
+marked as trusted is downgraded to the unverified pool and the counter and time
+is reset; unverified peers are simply removed.
+
+Connected peers will periodically be pinged, and their connection state
```

Comment:
Might help the reader to point out that the ping is solely for chain status information - i.e. it is not needed to track connection state etc.

Reply:
Will this change at some point? There was some discussion regarding connection timeout / cleanup regarding the system tests...

Reply:
A 2 minutes ping with retry would not be very helpfull for detecting lost connections anyway. If we wanted to speed up dead connection detection we would need a ping frequency a lot higher or a different kind of ping, maybe just sending a dummy byte and monitoring activity...